rfcs/2020-08-26-3191-host-metrics.md

```diff
@@ -0,0 +1,169 @@
+# RFC 3191 - 2020-08-26 - Collecting host-based metrics
+
+This RFC is to introduce a new metrics source to consume host-based metrics. The high level plan is to implement one (or more?) sources that collect CPU, disk, and memory metrics from Linux-based hosts.
+
+## Scope
+
+This RFC will cover:
+
+- A new source for host-based metrics, specifically:
+  - CPU
+  - Memory
+  - Disk
+- Collection on Linux, Windows, OSX.
+
+This RFC will not cover:
+
+- Other host metrics.
+- Other platforms
+
+## Motivation
+
+Users want to collect, transform, and forward metrics to better observe how their hosts are performing.
+
+## Internal Proposal
+
+Build a single source called `host_metrics` (name to be confirmed) to collect host/system level metrics.
+
+I've found a number of possible Rust-based solutions for implementing this collection, cross-platform.
+
+- https://crates.io/crates/heim (Rust)
+- https://lib.rs/crates/sys-info (Rust + C)
+- https://github.com/myfreeweb/systemstat (pure Rust)
+- https://docs.rs/sysinfo/0.3.19/sysinfo/index.html
+
+The Heim crate has a useful comparison doc showing the available tools and their capabilities.
+
+- https://github.com/heim-rs/heim/blob/master/COMPARISON.md
+
+For this implementation we're recommending the Heim crate based on platform and feature coverage.
+
+We'd use one of these to collect the following metrics:
+
+- `cpu_seconds_total` by mode (idle, nice, system, user) per CPU. (counter)
+- `disk_read_bytes_total` by disk (counter)
+- `disk_read_errors_total` by disk (counter)
+- `disk_read_retries_total` by disk (counter)
+- `disk_read_sectors_total` by disk (counter)
+- `disk_read_time_seconds_total` by disk (counter)
+- `disk_reads_completed_total` by disk (counter)
+- `disk_write_errors_total` by disk (counter)
+- `disk_write_retries_total` by disk (counter)
+- `disk_write_time_seconds_total` by disk (counter)
+- `disk_writes_completed_total` by disk (counter)
+- `disk_written_bytes_total` by disk (counter)
+- `disk_written_sectors_total` by disk (counter)
+- `filesystem_avail_bytes` labeled with device, filesystem type, and mountpoint (gauge)
+- `filesystem_device_error` labeled with device, filesystem type, and mountpoint (gauge)
+- `filesystem_total_file_nodes` labeled with device, filesystem type, and mountpoint (gauge)
+- `filesystem_free_file_nodes` labeled with device, filesystem type, and mountpoint (gauge)
+- `filesystem_free_bytes` labeled with device, filesystem type, and mountpoint (gauge)
+- `filesystem_size_bytes` labeled with device, filesystem type, and mountpoint (gauge)
```

Comment:
What is the distinction between `avail` and `free` or `size`? Also, there seems to be an inconsistency between avail/free/size for bytes, but free/total for file nodes.

Reply:
I've drawn these from Prometheus' approach specifically.

- `filesystem_avail_bytes` = Filesystem space available to non-root users in bytes.
- `filesystem_free_bytes` = Filesystem free space in bytes.
- `filesystem_size_bytes` = Filesystem size in bytes.

See https://www.robustperception.io/filesystem-metrics-from-the-node-exporter. I'll add an explainer to the RFC.

Reply:
My comment about the inconsistency was to suggest, for example, we could use avail/free/total for both:
```
filesystem_avail_bytes
filesystem_free_bytes
filesystem_total_bytes
filesystem_avail_file_nodes
filesystem_free_file_nodes
filesystem_total_file_nodes
```
Unless, of course, there is a preference for sticking to Prometheus' terminology.

Reply:
That's an interesting question - I replicated Prometheus' naming as a lot of folks are familiar with that. Also if folks have dashboards/alerts/etc that calculate things, like free space on a filesystem, then they don't need to rewrite them with new metric names.

Reply:
:thinking: I think the naming you (and prometheus) have feels natural to me. I'm used to referring to filesystems as having a size, probably because of my association with a physical disk and the fact that files themselves are referred to as having "sizes" (and not "total bytes"). Inodes, on the other hand feels more like a "discrete, countable" thing so total makes sense to me for the cap.

I can see an argument for consistency though.